addons/crm_partner_assign/models/crm_lead.py

```diff
@@ -0,0 +1,97 @@
+    partner_assigned_id = fields.Many2one('res.partner', string='Assigned Partner', track_visibility='onchange', help="Partner this case has been forwarded/assigned to.", index=True)
+    date_assign = fields.Date(string='Assignation Date', help="Last date this case was forwarded/assigned to a partner")
+
+    def get_interested_action(self, interested):
+        lead_channel_act = self.env.ref('crm_partner_assign.crm_lead_channel_interested_act')
+        action = lead_channel_act.read()[0]
+        action_context = eval(action['context'])
+        action_context['interested'] = interested
+        action['context'] = str(action_context)
+        return action
+    
+    @api.multi
+    def case_interested(self):
+        return self.get_interested_action(True)
+
+    @api.multi
+    def case_disinterested(self):
+        return self.get_interested_action(False)
+    
+    @api.multi
+    def assign_salesman_of_assigned_partner(self):
+        for lead in self.filtered(lambda x: x.partner_assigned_id.user_id != x.user_id and (x.probability > 0 and x.probability < 100) or x.active != False):
+            salesman_id = lead.partner_assigned_id.user_id.id
+            salesteam_id = self.on_change_user(salesman_id)['value'].get('team_id')
+            lead.write({'user_id': salesman_id, 'team_id': salesteam_id})
+
+    @api.multi
+    def _merge_data(self, oldest, fields):
+        fields += ['partner_latitude', 'partner_longitude', 'partner_assigned_id', 'date_assign']
+        return super(CrmLead, self)._merge_data(oldest, fields)
+ 
+    @api.onchange('partner_assigned_id')
+    def onchange_assign_id(self):
+        """This function updates the "assignation date" automatically, when manually assign a partner in the geo assign tab
+        """
+        self.date_assign = fields.Date.context_today(self)
+        self.user_id = self.partner_assigned_id.user_id.id
+
+    @api.multi
+    def action_assign_partner(self):
+        partner_dict = self.search_geo_partner()
+        for lead in self:
+            partner = partner_dict.get(lead.id, False)
+            if not partner:
+                continue
+            if not lead.partner_latitude or not lead.partner_longitude:
+                lead.assign_geo_localize()
+            if partner.user_id:
+                lead.allocate_salesman(partner.user_id.ids, team_id=partner.team_id.id)
+            lead.write({'date_assign': fields.Date.context_today(self), 
+                       'partner_assigned_id': partner.id})
+
+    @api.multi
+    def assign_geo_localize(self):
+        for lead in self:
+            result = geo_find(geo_query_address(street=lead.street,
+                                                zip=lead.zip,
+                                                city=lead.city,
+                                                state=lead.state_id.name,
+                                                country=lead.country_id.name))
```

Comment:
Don't you think we should not pass context here, to have country name always in english,
Check this: https://github.com/odoo/odoo/blob/master/addons/base_geolocalize/models/res_partner.py#L29
We specifically uses some english term here